org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/launch/BaseLaunchRequestJob.java

```diff
@@ -46,18 +47,29 @@ public Void runInToolingApi(CancellationTokenSource tokenSource, IProgressMonito
     }
 
     protected final void executeLaunch(CancellationTokenSource tokenSource, final IProgressMonitor monitor) throws Exception {
-        // todo (etst) close streams when done
-
-        BuildExecutionParticipants.activateParticipantPlugins();
+        BuildExecutionParticipants buildParticipants = BuildExecutionParticipants.create(CorePlugin.extensionManager().loadBuildExecutionParticipants());
+        
         monitor.beginTask(getJobTaskName(), IProgressMonitor.UNKNOWN);
 
         ProcessDescription processDescription = createProcessDescription();
         RunConfiguration runConfig = getRunConfig();
+        
         InternalGradleBuild gradleBuild = CorePlugin.internalGradleWorkspace().getGradleBuild(runConfig.getProjectConfiguration().getBuildConfiguration());
+
+        runConfig.getTasks().stream().forEach(
+            task -> {
+                Set<BuildExecutionParticipant> buildExecutionParticipant = buildParticipants.getBuildExecutionParticipant(task);
+
+                buildExecutionParticipant.stream().forEach(
+                    buildParticipant -> buildParticipant.preAction(runConfig.getProjectConfiguration().getProjectDir(), task, monitor)
```

Comment:
Do buildExecutionParticipants need to be called for each task?  Would it be good enough for each buildExecutionParticipant to just get the entire runConfig?  Or perhaps they should get the `gradleBuild` object?